taca/illumina/HiSeqX_Runs.py

```diff
@@ -63,102 +66,111 @@ def _copy_samplesheet(self):
-                with open(samplesheet_dest_10X, 'wb') as fcd:
-                    fcd.write(_generate_samplesheet_subset(self.runParserObj.samplesheet, self.lanes_10X, self.tenX_samples, is_10X = True))
-        else:
-            with chdir(self.run_dir):
-                samplesheet_dest="SampleSheet_0.csv"
-                with open(samplesheet_dest, 'wb') as fcd:
-                    if len(self.lanes_10X):
-                        fcd.write(_generate_samplesheet_subset(self.runParserObj.samplesheet, self.lanes_10X, self.tenX_samples, is_10X = True))
-                    else:
-                        fcd.write(_generate_samplesheet_subset(self.runParserObj.samplesheet, self.lanes_not_10X, self.tenX_samples, is_10X = False))
-
-        per_lane_base_masks = self._generate_per_lane_base_mask()
-        max_different_base_masks =  max([len(per_lane_base_masks[base_masks]) for base_masks in per_lane_base_masks])
-        if max_different_base_masks > 1:
-            # in a HiSeqX run I cannot have different index sizes in the SAME lane
-            logger.error("In FC {} found one or more lane with more than one base mask (i.e., different index sizes in \
-                         in the same lane".format(self.id))
-            return False
+        # Check sample types
+        sample_type_list = []
+        for lane, lane_contents in self.sample_table.items():
+            for sample in lane_contents:
+                if sample[sample.keys()[0]]['type'] not in sample_type_list:
+                    sample_type_list.append(sample[sample.keys()[0]]['type'])
+
+        # Go through sample_table for demultiplexing
         bcl2fastq_cmd_counter = 0
-        with chdir(self.run_dir):
-            # create Demultiplexing dir, this changes the status to IN_PROGRESS
-            if not os.path.exists("Demultiplexing"):
-                os.makedirs("Demultiplexing")
-        with chdir(self.run_dir):
-            if self.lanes_not_10X:
-               cmd_normal = self.generate_bcl_command(self.lanes_not_10X, bcl2fastq_cmd_counter)
-               misc.call_external_command_detached(cmd_normal, with_log_files = True, prefix="demux_{}".format(bcl2fastq_cmd_counter))
-               logger.info(("BCL to FASTQ conversion and demultiplexing started for "
-                   "normal run {} on {}".format(os.path.basename(self.id), datetime.now())))
-               bcl2fastq_cmd_counter += 1
-            if self.lanes_10X:
-               cmd_10X = self.generate_bcl_command(self.lanes_10X, bcl2fastq_cmd_counter, is_10X = True)
-               misc.call_external_command_detached(cmd_10X, with_log_files = True, prefix="demux_{}".format(bcl2fastq_cmd_counter))
-               logger.info(("BCL to FASTQ conversion and demultiplexing started for "
-                   "10X run {} on {}".format(os.path.basename(self.id), datetime.now())))
-               bcl2fastq_cmd_counter += 1
-        return True
+        for type in sample_type_list:
```

Comment:
Maybe consider renaming `type` so as not to confuse it with the built-in method `type()`?

Reply:
All `type` have been changed as `sample_type`. The changes have been tested and the code worked as expected.